examples/adrv9002-iiostream.c

```diff
@@ -0,0 +1,322 @@
+	}
+
+	/* printout some useful info */
+	ret = iio_channel_attr_read_longlong(chan, "rf_bandwidth", &val);
+	if (ret)
+		return ret;
+
+	info("adrv9002 bandwidth: %lld\n", val);
+
+	ret = iio_channel_attr_read_longlong(chan, "sampling_frequency", &val);
+	if (ret)
+		return ret;
+
+	info("adrv9002 sampling_frequency: %lld\n", val);
+
+	/* set the LO to 2.5GHz */
+	val = GHZ(2.5);
+	chan = iio_device_find_channel(phy, "altvoltage2", true);
+	if (!chan) {
+		error("Could not find TX LO channel\n");
+		return -ENODEV;
+	}
+
+	return iio_channel_attr_write_longlong(chan, "TX1_LO_frequency", val);
+}
+
+static void destroy_buffer(int status __notused, void *buffer)
+{
+	iio_buffer_destroy(buffer);
+}
+
+static void disable_channel(int status __notused, void *chan)
+{
+	iio_channel_disable(chan);
+}
+
+static void destroy_ctx(int status __notused, void *ctx)
+{
+	iio_context_destroy(ctx);
+}
+
+static int add_action_or_reset(void (*reset)(int , void *), void *arg)
+{
+	int ret;
+
+	ret = on_exit(reset, arg);
+	if (ret) {
+		error("Could not add exit action: %s\n", strerror(errno));
+		reset(ret, arg);
+	}
+
+	return ret;
+}
+
+static int stream_channels_get_enable(const struct iio_device *dev, struct iio_channel **chan,
+				      bool tx)
+{
+	int ret;
+	int c;
+	const char * const channels[] = {
+		"voltage0_i", "voltage0_q", "voltage0", "voltage1"
+	};
+
+	for (c = 0; c < 2; c++) {
+		const char *str = channels[tx * 2 + c];
+
+		chan[c] = iio_device_find_channel(dev, str, tx);
+		if (!chan[c]) {
+			error("Could not find %s channel tx=%d\n", str, tx);
+			return -ENODEV;
+		}
+
+		iio_channel_enable(chan[c]);
+		ret = add_action_or_reset(disable_channel, chan[c]);
+		if (ret)
+			return ret;
+
+	}
+
+	return 0;
+}
+/* Simplified version of the unaligned functions present in the linux kernel */
+static int16_t get_unaligned(const void *val)
+{
+	const struct {
+		 int16_t x;
+	} __packed *ptr = val;
+
+	return ptr->x;
+}
+
+static void put_unaligned(void *dest, int16_t val)
+{
+	struct {
+		 int16_t x;
+	} __packed *ptr = dest;
+
+	ptr->x = val;
+}
+
+static void stream(struct iio_buffer *rx, const struct iio_channel *rx_i_chan, ssize_t rx_sample,
+		   struct iio_buffer *tx, const struct iio_channel *tx_i_chan, ssize_t tx_sample)
+{
+	ssize_t nrx = 0;
+	ssize_t ntx = 0;
+
+	while (!stop) {
+		ssize_t nbytes_rx, nbytes_tx;
+		char *p_dat, *p_end;
+		ptrdiff_t p_inc;
+
+
+		nbytes_tx = iio_buffer_push(tx);
+		if (nbytes_tx < 0) {
+			error("Error pushing buf %zd\n", nbytes_tx);
+			return;
+		}
+
+		nbytes_rx = iio_buffer_refill(rx);
+		if (nbytes_rx < 0) {
+			error("Error refilling buf %zd\n", nbytes_rx);
+			return;
+		}
+
+		/* READ: Get pointers to RX buf and read IQ from RX buf port 0 */
+		p_inc = iio_buffer_step(rx);
+		p_end = iio_buffer_end(rx);
+		for (p_dat = iio_buffer_first(rx, rx_i_chan); p_dat < p_end; p_dat += p_inc) {
+			/* Example: swap I and Q */
+			int16_t i = get_unaligned(p_dat);
+			int16_t q = get_unaligned(&p_dat[2]);
```

Comment:
Why the unaligned access?

Reply:
I took this from [here](https://github.com/analogdevicesinc/libiio/blob/master/examples/adrv9009-iiostream.c#L269). On that code we might clearly have some misaligned access since we have no guarantees that p_dat sits on a address aligned with a short.  The compiler will also complain about that code (using `-Wcast-align`)...

If you do not want to have these helpers, plain `memcpy()` should also make the compiler happy. Or if you can guarantee (as you know the libiio internals better than me :)) that those addresses will always have some alignment guarantees, I can just use the cast as in the adrv9009 example (though I think that is a bad practice).

Reply:
The buffers and the samples are guaranteed to be aligned.

If you make `p_dat` a `int16_t *`, then the compiler won't complain and you can just read i/q without casting the pointer.

Reply:
Ok, that works as long as I also do the update like `p_dat += p_inc / 2`... I guess p_inc is the complete size of the enabled scan elements?

Reply:
Yes, `p_inc` is the sample size in bytes.
You can update with: `p_dat += p_inc / sizeof(*p_dat)` just so that it's obvious what it does.